File: homework/thinhbka/lesson5-graph/695_Max_Area_of_Island.py
# https://leetcode.com/problems/max-area-of-island/


DIR = [(0,1),(0,-1),(-1,0),(1,0)]
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ans = 0
        rows = len(grid)
        cols = len(grid[0])
        visited = [[False for _ in range(cols)] for _ in range(rows)]
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1:
                    frontier = [(i,j)]
                    cnt = 0
                    visited[i][j] = True
                    while frontier:
                        next = []
                        for (r,c) in frontier:
                            cnt+=1
                            for cur in DIR:
                                n_r = r+cur[0]
                                n_c = c+cur[1]
                                if 0<=n_r<rows and 0<=n_c < cols and visited[n_r][n_c] is False and grid[n_r][n_c]==1:
                                    visited[n_r][n_c] = True
                                    next.append((n_r,n_c))
                        frontier = next
                    ans = max(ans,cnt)
        return ans

File: homework/Truong_Duong/Lesson_7/695_Max_Area_of_Island.js
// Solution for https://leetcode.com/problems/max-area-of-island/

/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function (grid) {
    const R = grid.length, C = grid[0].length;
    const dx = [0, 0, -1, 1];
    const dy = [1, -1, 0, 0];

    const visited = {};
    for (let i = 0; i < R; i++) {
        visited[i] = new Array(C).fill(false);
    }

    const isValid = (r, c) => (r >= 0 && r < R && c >= 0 && c < C && grid[r][c] === 1);

    const dfs = (sr, sc) => {
        visited[sr][sc] = true;
        count += 1;

        for (let i = 0; i < 4; i++) {
            const vr = sr + dx[i];
            const vc = sc + dy[i];

            if (isValid(vr, vc) && !visited[vr][vc]) {
                dfs(vr, vc);
            }
        }
    }


    let ans = 0;
    for (let i = 0; i < R; i++) {
        for (let j = 0; j < C; j++) {
            if (grid[i][j] === 1 && !visited[i][j]) {
                var count = 0;
                dfs(i, j);
                ans = Math.max(ans, count);
            }
        }
    }
    return ans;
};
